cannon/mipsevm/tests/evm_multithreaded_test.go

```diff
@@ -1495,6 +1495,72 @@ func TestEVM_WakeupTraversal_Full(t *testing.T) {
 	}
 }
 
+func TestEVM_WakeupTraversal_WithExitedThreads(t *testing.T) {
+	var tracer *tracing.Hooks
+	addr := Word(0x1234)
+	wakeupVal := Word(0x999)
+	cases := []struct {
+		name              string
+		wakeupAddr        Word
+		futexAddr         Word
+		targetVal         Word
+		traverseRight     bool
+		activeStackSize   int
+		otherStackSize    int
+		exitedThreadIdx   []int
+		shouldClearWakeup bool
+		shouldPreempt     bool
+	}{
+		{name: "Wakeable thread exists among exited threads", wakeupAddr: addr, futexAddr: addr, targetVal: wakeupVal + 1, traverseRight: false, activeStackSize: 3, otherStackSize: 1, exitedThreadIdx: []int{1}, shouldClearWakeup: true, shouldPreempt: false},
+		{name: "All threads exited", wakeupAddr: addr, futexAddr: addr, targetVal: wakeupVal + 1, traverseRight: false, activeStackSize: 3, otherStackSize: 0, exitedThreadIdx: []int{0, 1, 2}, shouldClearWakeup: false, shouldPreempt: true},
+		{name: "Exited threads, no matching futex", wakeupAddr: addr, futexAddr: addr + 4, targetVal: wakeupVal, traverseRight: false, activeStackSize: 2, otherStackSize: 1, exitedThreadIdx: []int{0}, shouldClearWakeup: false, shouldPreempt: true},
+		{name: "Matching addr, not wakeable, with exited threads", wakeupAddr: addr, futexAddr: addr, targetVal: wakeupVal, traverseRight: true, activeStackSize: 3, otherStackSize: 0, exitedThreadIdx: []int{0, 1}, shouldClearWakeup: true, shouldPreempt: false},
+		{name: "Non-waiting threads with exited threads", wakeupAddr: addr, futexAddr: exec.FutexEmptyAddr, targetVal: 0, traverseRight: false, activeStackSize: 2, otherStackSize: 1, exitedThreadIdx: []int{0}, shouldClearWakeup: false, shouldPreempt: true},
+	}
+
+	for i, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			goVm, state, contracts := setup(t, i*1000, nil)
+			mttestutil.SetupThreads(int64(i*5000), state, c.traverseRight, c.activeStackSize, c.otherStackSize)
+			step := state.Step
+
+			state.Wakeup = c.wakeupAddr
+			state.GetMemory().SetWord(c.wakeupAddr&arch.AddressMask, wakeupVal)
+
+			threads := mttestutil.GetAllThreads(state)
```

Comment:
This setup that sets other threads to exited is OK. However, the active thread should always be exited since that's the specific scenario we're asserting here. Afterall, only a single state transition for the wakeup traversal occurs here.
Looking at the `exitedThreadIdx` test cases, I don't think this is always the case.

Reply:
Thanks, not sure if I understand correctly, I make a update in two ways in the new commit

1) Explicitly set the current active thread to exited state (`activeThread.Exited = true`) to test the wakeup traversal behavior when the active thread has exited.

2) Modified the test cases by resetting the `exitedThreadIdx` to make them more sensible.

Reply:
That's good. Could we also tweak other related state fields in the active thread? It'll be good to know that the VM behaves correctly for any `FutexAddr`, `FutexVal`, and `FutexTimeoutStep` value.